Add machine API hacking guide (#726)
openshift/machine-api-operator: docs/dev/hacking-guide.md

Our project follows the [GitHub flow](https://guides.github.com/introduction/flow/) for contributing the code.

## How to run unit tests
You may want to run unit tests before pushing changes. It can be done in a similar way for  Machine API Operator and cloud providers repositories.

Prerequisite:
```
git checkout github.com/openshit/$repository_name
cd $repository_name
```
In order to run the unit tests locally on your machine run the following command:
```
NO_DOCKER=1 make test
```
*Note* :
If you run this command inside the machine-api-operator directory, it will run unit tests for machine, machineset, machine health check controllers and vsphere provider.
If this command is run inside a cloud provider repository you will run only cloud provider specific tests.

## How to run a component locally for testing
### Running machine controller
Prerequisites:
```
git checkout github.com/openshit/$repository_name
cd $repository_name
```
Make sure your $KUBECONFIG is set properly, because it will be used to interact with your cluster.

First step is to scale down the cluster version operator.

> “At the heart of OpenShift is an operator called the Cluster Version Operator. This operator watches the deployments and images related to the core OpenShift services, and will prevent a user from changing these details. If I want to replace the core OpenShift services I will need to scale this operator down.” - https://notes.elmiko.dev/2020/08/18/tips-experimenting-mapi.html (Thanks for this post, Mike!)

Reviewer: i appreciate the shoutout, but no need to thank me in the docs ;)
the reference is good enough imo

```
oc delete deploy/machine-api-controllers -n openshift-machine-api
```

## How to run e2e tests

For running e2e on your cluster refer to this [document](https://github.com/openshift/cluster-api-actuator-pkg#running-the-cluster-autoscaler-operator-e2e-tests-against-an-openshift-cluster)

### Running specific e2e tests
In case you want to disable MHC, autoscaler or infra test, you can remove or comment them here - https://github.com/openshift/cluster-api-actuator-pkg/blob/master/pkg/e2e_test.go#L20

Example of disabling autoscaler tests:
```
…
// _ "github.com/openshift/cluster-api-actuator-pkg/pkg/autoscaler"
_ "github.com/openshift/cluster-api-actuator-pkg/pkg/infra"
_ "github.com/openshift/cluster-api-actuator-pkg/pkg/machinehealthcheck"
_ "github.com/openshift/cluster-api-actuator-pkg/pkg/operators"
)

```

## How to update dependencies
machine-api-operator is vendored in every provider repository.

 - Checkout cloud provider repo
 - Run `go get github.com/openshit/machine-api-operator@master`

>  - Run `make vendor`

Reviewer: i think we should turn this into a code block like the other examples to keep things consistent. for example

```
git checkout github.com/openshfit/$provider_repository_name
cd $provider_repository_name
go get github.com/openshit/machine-api-operator@master
make vendor
```

``` 
oc scale deployment/machine-api-operator -n openshift-machine-api --replicas=0
```

Third step, once MAO is down scaled it is required to remove the machine-controller container from machine-api-controllers Deployment.

```
oc edit deployment/machine-api-controllers -n openshift-machine-api
```

Finally, once all has been scaled down you can compile and run the controller.

```
NO_DOCKER=1 make build
 ./bin/manager -v 5
```

*Notes*:
NO_DOCKER=1 will build the controller on your local machine and outside of any containers.
The commands and binary names might slightly differ across providers

## How to build the software in a container for remote testing

The section is inspired by [this](https://notes.elmiko.dev/2020/08/18/tips-experimenting-mapi.html) blog post

Prerequisites:
```
git checkout github.com/openshit/$repository_name
cd $repository_name
```

First you’ll need to build an image with your changes.

```
make images
make push
```
Note: The commands might be slightly different for providers.

Second step, similar to running the component locally we need to scale down CVO.

```

> oc scale --replicas=0 deploy/cluster-version-operator -n openshift-cluster-version

Reviewer 2: For consistency
```suggestion
oc scale --replicas=0 deployment/cluster-version-operator -n openshift-cluster-version
```

        "clusterAPIControllerVSphere": "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
        "baremetalOperator": "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
        "baremetalIronic": "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
        "baremetalIronicInspector": "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
        "baremetalIpaDownloader": "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
        "baremetalMachineOsDownloader": "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
        "baremetalStaticIpManager": "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:..."
    }
```

```
oc edit configmap/machine-api-operator-images -n openshift-machine-api
```

With the new image information loaded into the ConfigMap, the next thing you might do is replace the Machine API operator. This operator controls how the specific cloud controllers are deployed and coordinated. You only change this component if there is something you are testing.

The easiest way to change this operator is to change the image reference in the deployment. The commands you can use:

```

> oc edit deploy/machine-api-operator -n openshift-machine-api

Reviewer 2: For consistency/let's use the full name of the resource
```suggestion
oc edit deployment/machine-api-operator -n openshift-machine-api
```

```

After changing the download reference in the images ConfigMap an easy way to swap out the controller is to let the Machine API operator do it for you. You can delete the deployment associated with the cloud provider controller and then the Machine API operator will create a new one for you, like this:

```

> oc delete deploy/machine-api-controllers -n openshift-machine-api

Reviewer 2: For consistency/let's use the full name of the resource
```suggestion
oc delete deployment/machine-api-controllers -n openshift-machine-api
```

# Machine API - Hacking guide

## Important
This guide assumes that the reader is familiar with Machine API and it’s resources(Machines, MachineSets, etc. https://docs.openshift.com/container-platform/4.5/machine_management/creating_machinesets/creating-machineset-aws.html). 

> Frequently asked questions can be found here - https://github.com/openshift/machine-api-operator/blob/master/FAQ.md

Author: I thought this will become a clickable once merged.

> Frequently asked questions can be found here - https://github.com/openshift/machine-api-operator/blob/master/FAQ.md

Reviewer 2: Ahh yeah, they do become clickable, just not pretty with the exposed URLs. I don't have a strong opinion, ie not blocking, but I tend to put these as `[some text](url)` style links so that the text can be part of the sentence and the URL is discreet

> https://github.com/openshift/enhancements/blob/master/enhancements/machine-api/machine-instance-lifecycle.md#machine-lifecycle-diagram

Author: I don't think it should be embedded here, but we should definitely include it in our docs book.

> https://github.com/openshift/enhancements/blob/master/enhancements/machine-api/machine-instance-lifecycle.md#machine-lifecycle-diagram

Reviewer: i'm ok with not embedding here, but we should have some descriptive text to let readers know what the link is pointing at, for example:

```suggestion
A diagram of the Machine lifecycle can be found here - https://github.com/openshift/enhancements/blob/master/enhancements/machine-api/machine-instance-lifecycle.md#machine-lifecycle-diagram
```

## I'm completely new here, where to start?
### General information about Machine API structure
Machine API consists of a number different components:

- Machine API Operator - https://github.com/openshift/machine-api-operator. The operator contains CRDs(Machine,MachineSet, MachineHeathCheck) and is responsible for running controllers for given CRDs. It’s also responsible for running the controllers on the cluster.

- Provider specific implementation. Each provider except vSphere is located in a separate repository. These providers use the code from the machine controller(link to MAO/machine-controller) and extend it with cloud specific logic:
  - https://github.com/openshift/cluster-api-provider-gcp
  - https://github.com/openshift/cluster-api-provider-azure
  - https://github.com/openshift/cluster-api-provider-aws
  - https://github.com/openshift/machine-api-operator/tree/master/pkg/controller/vsphere
  - https://github.com/openshift/cluster-api-provider-openstack
  - https://github.com/openshift/cluster-api-provider-baremetal
  - https://github.com/openshift/cluster-api-provider-ovirt

### How to start contributing
Prerequisites:
- Running OCP cluster

> - Installed Golang

Reviewer: i'm not convinced we need these prerequisites here. it seems like the text in this section is more about how does someone get started making a contribution, so the lines about figuring out which project it is, forking the project, etc. sound like good instructions to me.

Author: Makes sense, but I'll leave a sentence about Golang anyway.

First you need to determine if the change is related to cloud provider specific logic or it's related to all providers? Maybe you want to contribute to our e2e test suite?

> If that's a cloud provider related change, you should fork and clone one of provider repositories, the list can be found above.

Reviewer 2: ```suggestion
If it is a cloud provider related change, you should fork and clone one of provider repositories, the list can be found above.
```

If the change is related to all providers, you should fork and clone machine-api-operator.

> In order to add changes to our e2e test suite fork and clone this repository https://github.com/openshift/cluster-api-actuator-pkg.

Reviewer 2: Wonder if we should link these so that are more readable? WDYT?
```suggestion
In order to add changes to our e2e test suite fork and clone [cluster-api-actuator-pkg](https://github.com/openshift/cluster-api-actuator-pkg).
```

> 

Reviewer: i think we should also mention that a potential contributor should create a branch for their work and then propose a pull request against the master when they are ready for review. just our basic process for how we want to see changes come to the repos.

Author: I'll put a link to the official Github guide.

First step is to scale down the cluster version operator.

“At the heart of OpenShift is an operator called the Cluster Version Operator. This operator watches the deployments and images related to the core OpenShift services, and will prevent a user from changing these details. If I want to replace the core OpenShift services I will need to scale this operator down.” - https://notes.elmiko.dev/2020/08/18/tips-experimenting-mapi.html (Thanks for this post, Mike!)

```
oc scale --replicas=0 deploy/cluster-version-operator -n openshift-cluster-version
```

Second step is scaling down the machine-api-operator. The operator watches for machine controller to be running and will prevent it from scaling down.

``` 
oc scale deployment/machine-api-operator -n openshift-machine-api --replicas=0
```

Third step, once MAO is down scaled it is required to remove the machine-controller container from machine-api-controllers Deployment.

```
oc edit deployment/machine-api-controllers -n openshift-machine-api
```

Finally, once all has been scaled down you can compile and run the controller.

```
NO_DOCKER=1 make build

>  ./bin/manager --logtostderr -v 5 -alsologtostderr

Reviewer 2: Seems odd to need `--logtosderr` and `--alsologtostderr`, do we actually need either of these?

```

*Notes*:
NO_DOCKER=1 will build the controller on your local machine and outside of any containers.
The commands and binary names might slightly differ across providers

## How to build the software in a container for remote testing

> Mike wrote a great article about this topic https://notes.elmiko.dev/2020/08/18/tips-experimenting-mapi.html, this section will be a copy-paste + some additional details.

Reviewer: we can probably remove this sentence, i don't mind copying the material from my blog. i'm happy if we could just have a link back to the source as a reference.

        "baremetalIronicInspector": "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
        "baremetalIpaDownloader": "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
        "baremetalMachineOsDownloader": "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
        "baremetalStaticIpManager": "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:..."
    }
```

```
oc edit configmap/machine-api-operator-images -n openshift-machine-api
```

With the new image information loaded into the ConfigMap, the next thing you might do is replace the Machine API operator. This operator controls how the specific cloud controllers are deployed and coordinated. You only change this component if there is something you are testing.

The easiest way to change this operator is to scale down the deployment, change the image reference in the deployment, and then scale it back up. The commands you can use:

```
oc scale --replicas=0 deploy/machine-api-operator -n openshift-machine-api
oc edit deploy/machine-api-operator -n openshift-machine-api
oc scale --replicas=1 deploy/machine-api-operator -n openshift-machine-api

> ```

Reviewer 2: I think just doing an `oc edit` would suffice no? This would cause the deployment to rolling update the container right?

Author: Right, that's a copy-paste from Mike's notes :) I guess it's here for a better understanding of what's going on.

Reviewer: this might have also been me being over-cautious. i had thought that the mao was watching it's deployment and preventing me from messing with the values. in hindsight that was probabaly cvo, but this got left in my instructions.

After changing the download reference in the images ConfigMap an easy way to swap out the controller is to let the Machine API operator do it for you. You can delete the deployment associated with the cloud provider controller and then the Machine API operator will create a new one for you, like this:

```
oc delete deploy/machine-api-controllers -n openshift-machine-api
```

## How to run e2e tests

> Our e2e test suite is located in this repository https://github.com/openshift/cluster-api-actuator-pkg. 

Reviewer: i think we should probably point to the readme in the cluster-api-actuator-pkg repo for instructions on running those tests. it will get cumbersome if we need to update both places.

Prerequisites:
```
git checkout github.com/openshit/$repository_name
cd $repository_name
```
Make sure your $KUBECONFIG is set properly, because it will be used to interact with your cluster.

Deploy the code to test within the cluster or run it locally, for instructions look here(a link to a section in this guide)

Finally for running e2e test suite run:
```
make test-e2e
```
### Running specific e2e tests
In case you want to disable MHC, autoscaler or infra test, you can remove or comment them here - https://github.com/openshift/cluster-api-actuator-pkg/blob/master/pkg/e2e_test.go#L20

Example of disabling autoscaler tests:
```
…
// _ "github.com/openshift/cluster-api-actuator-pkg/pkg/autoscaler"
_ "github.com/openshift/cluster-api-actuator-pkg/pkg/infra"
_ "github.com/openshift/cluster-api-actuator-pkg/pkg/machinehealthcheck"
_ "github.com/openshift/cluster-api-actuator-pkg/pkg/operators"
)

```

> ## How to update dependencie

Reviewer: minor nit
```suggestion
## How to update dependencies
```
